daemon/mqtt.c

```diff
@@ -0,0 +1,401 @@
+                usleep(10000 * 1000);
+            } else
+                error("Loop error code %d (%s)", rc, mosquitto_strerror(rc));
+            continue;
+        }
+
+//        if (unlikely(rc == MOSQ_ERR_INVAL)) {
+//            // TODO: handle the edge case -- reset the connection and retry (for now sleep and loop)
+//            error("Invalid MQTT connection; dying?");
+//            break;
+//        }
+
+        // TODO: handle connection lost and REINIT gracefully -- for now handles internally
+        // TODO: need to decide retry interval and total duration
+//        if (rc == MOSQ_ERR_CONN_LOST) {
+//            mqtt_init(MQTT_REINIT);
+//        }
+
+    } // forever
+    mqtt_shutdown();
+
+    netdata_thread_cleanup_pop(1);
+    return NULL;
+}
+
+//    rc = mosquitto_publish(mosq, NULL, "netdata/alarm", strlen(mqtt_message), mqtt_message, 0, 0);
+
+/*
+ * Send a message to the cloud, using a base topic and sib_topic
+ * The final topic will be in the form <base_topic>/<sub_topic>
+ * If base_topic is missing then the global_base_topic will be used (if available)
+ *
+ */
+int mqtt_send(char *base_topic, char *sub_topic, char *message)
+{
+    int rc;
+    static int skip_due_to_shutdown = 0;
+    static char *global_base_topic = NULL;
+    char topic[NETDATA_MQTT_MAX_TOPIC + 1];
+    char *final_topic;
+
+    if (!am_i_claimed())
+        return 0;
+
+    if (unlikely(netdata_exit)) {
+
+        if (unlikely(!mqtt_connection_initialized))
+            return 0;
+
+        ++skip_due_to_shutdown;
+        if (unlikely(!(skip_due_to_shutdown % 100)))
+            info("%d messages not sent -- shutdown in progress", skip_due_to_shutdown);
+        return 0;
+    }
+
+    if (unlikely(!message))
+        return 0;
+
+    // Just wait to be initialized
+    if (unlikely(!mqtt_connection_initialized)) {
+        time_t now = now_realtime_sec();
+
+        while (!mqtt_connection_initialized && (now_realtime_sec() - now) < NETDATA_MQTT_INITIALIZATION_WAIT) {
+            usleep(NETDATA_MQTT_INITIALIZATION_SLEEP_WAIT * 1000);
+        }
+
+        if (unlikely(!mqtt_connection_initialized)) {
+            error("MQTT connection not active");
+            return 0;
+        }
+    }
+
+    if (unlikely(!global_base_topic))
+        global_base_topic = GET_PUBLISH_BASE_TOPIC;
+
+    if (unlikely(!base_topic)) {
+        if (unlikely(!global_base_topic))
+            final_topic = sub_topic;
+        else {
+            snprintfz(topic, NETDATA_MQTT_MAX_TOPIC, "%s/%s", global_base_topic, sub_topic);
+            final_topic = topic;
+        }
+    } else {
+        snprintfz(topic, NETDATA_MQTT_MAX_TOPIC, "%s/%s", base_topic, sub_topic);
+        final_topic = topic;
+    }
+
+    rc = mosquitto_pub_topic_check(final_topic);
+    if (unlikely(rc != MOSQ_ERR_SUCCESS))
+        return rc;
+
+    int msg_len = strlen(message);
+
+    // TODO: handle encoding validation -- the message should be UFT8 encoded by the sender
```

Comment:
I think we can skip this - the protocol just defines a payload as an application specific stream of bytes.

Reply:
yes, may not be needed..  just put a note there in case we need it